bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
+	matches := Signal{}
+	descriptions := []string{}
+
+	if len(s.Branches) == 0 {
+		logger.Debug().Msgf("No branch singals have been provided to match against")
+		return descriptions, matches, nil
 	}
+
+	targetBranch, _ := pullCtx.Branches()
+
 	for _, signalBranch := range s.Branches {
 		if targetBranch == signalBranch {
-			return true, fmt.Sprintf("pull request target is a %s branch: %q", tag, signalBranch), nil
+			matches = append(matches, signalBranch)
+			descriptions = append(descriptions, fmt.Sprintf("pull request target is a %s branch: %q", tag, signalBranch))
 		}
 	}
+
+	return descriptions, matches, nil
+}
+
+// branchPatternMatches determines which branch pattern signals match the given PR. It returns:
+// - A list of descriptions for each matched signal
+//   - These will only include the first item the signal matched when there is more than one
+// - A list of the matched signals
+func (s *Signals) branchPatternMatches(ctx context.Context, pullCtx pull.Context, tag string) ([]string, Signal, error) {
+	logger := zerolog.Ctx(ctx)
+	matches := Signal{}
+	descriptions := []string{}
+
+	if len(s.BranchPatterns) == 0 {
+		logger.Debug().Msgf("No branch pattern singals have been provided to match against")
+		return descriptions, matches, nil
+	}
+
+	targetBranch, _ := pullCtx.Branches()
+
 	for _, signalBranch := range s.BranchPatterns {
 		if matched, _ := regexp.MatchString(fmt.Sprintf("^%s$", signalBranch), targetBranch); matched {
-			return true, fmt.Sprintf("pull request target branch (%q) matches pattern: %q", targetBranch, signalBranch), nil
+			matches = append(matches, signalBranch)
+			descriptions = append(descriptions, fmt.Sprintf("pull request target branch (%q) matches pattern: %q", targetBranch, signalBranch))
 		}
 	}
 
-	return false, fmt.Sprintf("pull request does not match the %s", tag), nil
+	return descriptions, matches, nil
+}
+
+// maxCommitsMatches determines if the number of commits in a PR is at or below a given max. It returns:
+// - An empty list if there is no match, otherwise a single string description of the match
+// - A match value of 0 if there is no match, otherwise the value of the max commits signal
+func (s *Signals) maxCommitsMatches(ctx context.Context, pullCtx pull.Context, tag string) ([]string, int, error) {
+	logger := zerolog.Ctx(ctx)
+	matches := 0
+	descriptions := []string{}
+
+	if s.MaxCommits <= 0 {
+		logger.Debug().Msgf("No valid max commits value has been provided to match against")
+		return descriptions, matches, nil
+	}
+	
+	commits, _ := pullCtx.Commits(ctx)
+
+	if len(commits) < s.MaxCommits {
+		matches = s.MaxCommits
+		descriptions = append(descriptions, fmt.Sprintf("pull request has %q commits, which is less than the maximum of %q", len(commits), s.MaxCommits))
+	}
+
+	if len(commits) == s.MaxCommits {
+		matches = s.MaxCommits
+		descriptions = append(descriptions, fmt.Sprintf("pull request has %q commits, which is the maximum", len(commits)))
+	}
```

Comment:
Does this need to be a separate condition? It seems like we could write a single description that gives the necessary information in both cases.